flows/resumes/base.go

```diff
@@ -0,0 +1,92 @@
+package resumes
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/nyaruka/goflow/flows"
+	"github.com/nyaruka/goflow/flows/events"
+	"github.com/nyaruka/goflow/utils"
+)
+
+type readFunc func(session flows.Session, data json.RawMessage) (flows.Resume, error)
+
+var registeredTypes = map[string]readFunc{}
+
+// RegisterType registers a new type of trigger
+func RegisterType(name string, f readFunc) {
+	registeredTypes[name] = f
+}
+
+type baseResume struct {
+	environment utils.Environment
+	contact     *flows.Contact
+	resumedOn   time.Time
+}
+
+func newBaseResume(env utils.Environment, contact *flows.Contact) baseResume {
+	return baseResume{environment: env, contact: contact, resumedOn: utils.Now()}
+}
+
+func (r *baseResume) Environment() utils.Environment { return r.environment }
+func (r *baseResume) Contact() *flows.Contact        { return r.contact }
+func (r *baseResume) ResumedOn() time.Time           { return r.resumedOn }
+
+// Apply applies our state changes and saves any events to the run
+func (r *baseResume) Apply(run flows.FlowRun, step flows.Step) error {
+	if r.environment != nil {
+		run.Session().SetEnvironment(r.environment)
+
+		// TODO diffing
+
+		run.LogEvent(step, events.NewEnvironmentChangedEvent(r.Environment()))
+	}
+	if r.contact != nil {
+		run.SetContact(r.contact)
+		run.Session().SetContact(r.contact)
+
+		// TODO diffing
```

Comment:
still pondering on giving these things complex .Equal() methods or relying on string comparison of their JSON forms

Reply:
Ya, its a bit of work though diffing would lead to clearer (and smaller!) logs so may be worth it.